File: Learning_Japan_project_fe/learning-app-fe/src/services/lessonService.ts
import { http } from "@/lib/http";
import { API_ENDPOINTS } from "@/config/api";
import { LessonLevel } from "@/enums/LessonLevel";

export interface CreateLessonRequest {
  sectionId: string;
  title: string;
  lessonOrder: number;
}

export interface LessonResponse {
  id: string;
  title: string;
  lessonOrder: number;
  createdAt: string;
}

/* ===================== SERVICE ===================== */

export const lessonService = {
  /**
   * ✅ Create lesson
   */
  async create(request: CreateLessonRequest): Promise<string> {
    const res = await http.post<any>(API_ENDPOINTS.LESSON.CREATE, request);
    return typeof res === 'object' && res !== null ? res.id : String(res);
  },
  update: async (id: string, data: CreateLessonRequest): Promise<string> => {
    const res = await http.put<any>(API_ENDPOINTS.LESSON.UPDATE(id), data);
    return typeof res === "object" && res !== null ? res.id : String(res);
  },
  /**
   * ✅ Get lessons by section
   */
  getBySection(sectionId: string): Promise<LessonResponse[]> {
    return http.get<LessonResponse[]>(
      API_ENDPOINTS.LESSON.GET_BY_SECTION(sectionId)
    );
  },

  /**
   * ✅ Get lesson detail
   */
  getDetail(lessonId: string): Promise<LessonResponse> {
    return http.get<LessonResponse>(API_ENDPOINTS.LESSON.GET_DETAIL(lessonId));
  },

  /**
   * ✅ Delete lesson
   */
  delete(lessonId: string): Promise<void> {
    return http.delete<void>(API_ENDPOINTS.LESSON.DELETE(lessonId));
  },
};

File: learningApp/src/main/java/com/example/learningApp/service/course/section/lesson/LessonService.java
package com.example.learningApp.service.course.section.lesson;

import com.example.learningApp.common.EntityFinder;
import com.example.learningApp.dto.request.course.section.lesson.CreateLessonRequest;
import com.example.learningApp.dto.response.lesson.LessonResponse;
import com.example.learningApp.entity.Lesson;
import com.example.learningApp.entity.Section;
import com.example.learningApp.mapper.LessonMapper;
import com.example.learningApp.repository.LessonRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class LessonService {

    private final LessonRepository lessonRepository;
    private final EntityFinder finder;
    private final LessonMapper lessonMapper;

    @Transactional
    public String createLesson(CreateLessonRequest request) {

        Section section = finder.sectionId(request.getSectionId());
        Lesson lesson = lessonMapper.toLesson(request);
        lesson.setSection(section);
        lesson.setCreatedAt(LocalDateTime.now());
        lessonRepository.save(lesson);

        return "Create lesson successfully";
    }


    /* ===================== GET BY SECTION ===================== */

    @Transactional(readOnly = true)
    public List<LessonResponse> getLessonsBySection(String sectionId) {

        return lessonRepository.findBySectionIdOrderByLessonOrderAsc(sectionId)
                .stream()
                .map(lessonMapper::toLessonResponse)
                .toList();
    }

    /* ===================== GET DETAIL ===================== */

    @Transactional(readOnly = true)
    public LessonResponse getLessonDetail(String lessonId) {

        Lesson lesson = finder.lessonId(lessonId);
        return lessonMapper.toLessonResponse(lesson);
    }
    @Transactional
    public String updateLesson(String lessonId, CreateLessonRequest request) {

        Lesson lesson = finder.lessonId(lessonId);

        // Nếu cho phép đổi section
        if (request.getSectionId() != null) {
            Section section = finder.sectionId(request.getSectionId());
            lesson.setSection(section);
        }

        lesson.setTitle(request.getTitle());
        lesson.setDescription(request.getDescription());
        lesson.setLessonOrder(request.getLessonOrder());

        lessonRepository.save(lesson);

        return "Update lesson successfully";
    }
    /* ===================== DELETE ===================== */

    @Transactional
    public void deleteLesson(String lessonId) {

        Lesson lesson = finder.lessonId(lessonId);
        lessonRepository.delete(lesson);
    }
}

